Source/FieldSolver/FiniteDifferenceSolver/HybridPICModel/ExternalVectorPotential.cpp

```diff
@@ -0,0 +1,376 @@
+                    warpx.m_fields.get(Aext_field, Direction{2}, lev),
+                    "A", "z");
+#else
+                warpx.ReadExternalFieldFromFile(m_external_file_path[i],
+                    warpx.m_fields.get(Aext_field, Direction{0}, lev),
+                    "A", "x");
+                warpx.ReadExternalFieldFromFile(m_external_file_path[i],
+                    warpx.m_fields.get(Aext_field, Direction{1}, lev),
+                    "A", "y");
+                warpx.ReadExternalFieldFromFile(m_external_file_path[i],
+                    warpx.m_fields.get(Aext_field, Direction{2}, lev),
+                    "A", "z");
+#endif
+            }
+        } else {
+            // Initialize the A fields from expression
+            m_A_external_parser[i][0] = std::make_unique<amrex::Parser>(
+                utils::parser::makeParser(m_Ax_ext_grid_function[i],{"x","y","z","t"}));
+            m_A_external_parser[i][1] = std::make_unique<amrex::Parser>(
+                utils::parser::makeParser(m_Ay_ext_grid_function[i],{"x","y","z","t"}));
+            m_A_external_parser[i][2] = std::make_unique<amrex::Parser>(
+                utils::parser::makeParser(m_Az_ext_grid_function[i],{"x","y","z","t"}));
+            m_A_external[i][0] = m_A_external_parser[i][0]->compile<4>();
+            m_A_external[i][1] = m_A_external_parser[i][1]->compile<4>();
+            m_A_external[i][2] = m_A_external_parser[i][2]->compile<4>();
+
+            // check if the external current parsers depend on time
+            for (int idim=0; idim<3; idim++) {
+                const std::set<std::string> A_ext_symbols = m_A_external_parser[i][idim]->symbols();
+                WARPX_ALWAYS_ASSERT_WITH_MESSAGE(A_ext_symbols.count("t") == 0,
+                    "Externally Applied Vector potential time variation must be set with A_time_external_function(t)");
+            }
+
+            // Initialize data onto grid
+            for (auto lev = 0; lev <= warpx.finestLevel(); ++lev) {
+                warpx.ComputeExternalFieldOnGridUsingParser(
+                    Aext_field,
+                    m_A_external[i][0],
+                    m_A_external[i][1],
+                    m_A_external[i][2],
+                    lev, PatchType::fine,
+                    warpx.GetEBUpdateEFlag(),
+                    false);
+
+                for (int idir = 0; idir < 3; ++idir) {
+                    warpx.m_fields.get(Aext_field, Direction{idir}, lev)->
+                        FillBoundary(warpx.Geom(lev).periodicity());
+                }
+            }
+        }
+
+        amrex::Gpu::streamSynchronize();
+
+        CalculateExternalCurlA(m_field_names[i]);
+
+        // Generate parser for time function
+        m_A_external_time_parser[i] = std::make_unique<amrex::Parser>(
+            utils::parser::makeParser(m_A_ext_time_function[i],{"t",}));
+        m_A_time_scale[i] = m_A_external_time_parser[i]->compile<1>();
```

Comment:
It occurred to me that it would be good to add a bit of logic here that will check whether a time dependence was specified, and if so do a divergence cleaning on the specified vector potential, since $E = \partial A/\partial t$ requires the Coulomb gauge.
What do you think? If not we should probably at least print a warning to users that it is expected that A is divergence free.

Reply:
I will add a warning here and then plug in the div cleaner as a follow up PR. I will need to test it to make sure it still works as expected with the offset staggering of A vs B.